chapter5/src/distance_test.go

```diff
@@ -0,0 +1,125 @@
+package chapter5
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDistance_CalculateDistance(t *testing.T) {
+	oxford := CreateCity(51.45, 1.15)
+	turin := CreateCity(45.04, 7.42)
+	ml, km := CalculateHarvesineDistance(oxford, turin)
+	if ml < 0.0 || km < 0.0 {
+		t.Errorf("The value cant be less than zero")
+	}
+}
+
+func TestDistance_CalculateDistance_ManualBetweenPoints(t *testing.T) {
+	oxford := CreateCity(51.45, 1.15)
+	turin := CreateCity(45.04, 7.42)
+	calculateOxford, _ := oxford.CalculateHarvesineDistanceManual(turin)
+	calculateTuron, _ := turin.CalculateHarvesineDistanceManual((oxford))
+	assert.Equal(t, calculateOxford, calculateTuron, "The comparison is the same")
+}
+
+func TestDistance_FullTest_Athens_Amsterdam(t *testing.T) {
+	type args struct {
+		athens    City
+		amsterdam City
+	}
+	tests := map[string]struct {
+		args        args
+		want        float64
+		expectedErr string
+	}{
+		"success": {args: args{athens: City{37.983972, 23.727806}, amsterdam: City{52.366667, 4.9}}, want: 2163.2310285824487},
+		"failure": {args: args{athens: City{0.0, 0.0}, amsterdam: City{0.0, 0.0}}, expectedErr: "The value cant be zero"},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			got, error := tt.args.athens.CalculateHarvesineDistanceManual(&tt.args.amsterdam)
+			if tt.expectedErr != "" {
+				assert.EqualError(t, error, tt.expectedErr)
+				assert.Equal(t, 0.0, got)
+			} else {
+				assert.NoError(t, error)
+				assert.Equal(t, tt.want, got)
```

Comment:
Just like with other programming languages, you want to avoid testing equality between two floating point numbers as this can lead to unexpected results. Use this function instead: https://pkg.go.dev/github.com/stretchr/testify/assert#InDelta

Reply:
Good point